core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1392,17 +1378,7 @@ class DAGScheduler(
 
       if (filesLost || !env.blockManager.externalShuffleServiceEnabled) {
         logInfo("Shuffle files lost for executor: %s (epoch %d)".format(execId, currentEpoch))
-        // TODO: This will be really slow if we keep accumulating shuffle map stages
-        for ((shuffleId, stage) <- shuffleIdToMapStage) {
-          stage.removeOutputsOnExecutor(execId)
-          mapOutputTracker.registerMapOutputs(
-            shuffleId,
-            stage.outputLocInMapOutputTrackerFormat(),
```

Comment:
This was a potential inefficiency in the old code. If you lost a single executor then we'd end up overwriting the `MapOutputTracker`'s state for every ShuffleMapStage. We'd have to iterate through each task once in  `stage.removeOutputsOnExecutor` and then a second time in `stage.outputLocInMapOutputTrackerFormat()` and in both iterations we'd have to iterate on a `List` for each task too. 

While the new code still needs to scan every task it doesn't have to scan a separate list per task and only needs a single scan, not two.